lib/widget/paragraph.cpp

```diff
@@ -17,147 +17,495 @@
+            partialWordFront->offset = currentLineBack.offset;
+        }
+
+        for (auto fragment: partialWord)
+        {
+            currentLine.push_back(fragment);
+        }
+
+        nextOffset += partialWordWidth;
+        partialWordWidth = 0;
+        partialWord.clear();
+    }
+
+    void pushFragment(FlowLayoutElementDescriptor const &elementDescriptor, size_t begin, size_t end)
+    {
+        auto width = elementDescriptor.getWidth(begin, end - begin);
+        partialWord.push_back({currentElementId, begin, end - begin, width, nextOffset + partialWordWidth});
+        partialWordWidth += width;
+    }
+
+    void placeLine(FlowLayoutElementDescriptor const &elementDescriptor, size_t begin, size_t end)
+    {
+        auto current = begin;
+
+        while (current < end)
+        {
+            long fragmentFits;
+
+            if (nextOffset + partialWordWidth + elementDescriptor.getWidth(current, end - current) > maxWidth)
+            // fragment doesn't fit completely in the current line
+            {
+                fragmentFits = current - 1;
+                size_t fragmentDoesntFit = end;
+                while (fragmentDoesntFit - fragmentFits > 1)
+                {
+                    auto middle = (fragmentFits + fragmentDoesntFit) / 2;
+                    if (nextOffset + partialWordWidth + elementDescriptor.getWidth(current, middle - current) > maxWidth)
+                    {
+                        fragmentDoesntFit = middle;
+                    } else {
+                        fragmentFits = middle;
+                    }
+                }
+            }
+            else
+            {
+                fragmentFits = end;
+            }
+
+            auto fragmentEnd = fragmentFits + 1;
+            while ((fragmentEnd >= elementDescriptor.size() || !elementDescriptor.isWhitespace(fragmentEnd - 1)) && fragmentEnd > current)
+            {
+                fragmentEnd--;
+            }
+
+            if (fragmentEnd > current)
+            // the fragment ending with a whitespace fits within the line
+            {
+                pushFragment(elementDescriptor, current, fragmentEnd - 1);
+                endWord();
+                nextOffset += elementDescriptor.getWidth(fragmentEnd - 1, 1);
+                current = fragmentEnd;
+            }
+            else if (fragmentFits == end)
+            // the fragment is a single word, and fits in the line,
+            // but it doesn't end in whitespace and the next element might be part of this word
+            {
+                pushFragment(elementDescriptor, current, fragmentFits);
+                current = fragmentFits;
+            }
+            else if (nextOffset > 0)
+            // word doesn't fit in the current line
+            {
+                 breakLine();
+            }
+            else
+            // word doesn't fit in an empty line
+            {
+                auto end = std::max((size_t)fragmentFits, current + 1);
```

Comment:
This line is yielding:
`Local variable 'end' hides a parameter of the same name.`